chore: code style change, lint using modified eslint-config-airbnb

vvo opened the pull request:
Following #128 and various discussions we already had on code style, I
configured our lint on airbnb/javascript style guide using eslint.

I chose to stay with ESLint because:
- it's becoming a good linting standard and has the best support for both ES6,
  jsx, react..
- already what we know, do not have to change again
- other alternatives were not as good as what we wanted

Not a single linter has a "Automatically reformat this whole code
option, do it well". ESLint has some wishes for ESLint 2.0 but may be
later.

In the end we will struggle a little bit on following our own Algolia
style at the begining but it will be worth it soon as we will be more
and more JavaScript`er. I recall @bobylito wanting to do so early on
when we both joined. Well, now it's time!

The goal of this PR is to provide a solid ground for our linting rules
amongst all our JS projects.

The currently picked coding style is the one from Airbnb, with
modifications.

You can find the modified rules here: https://github.com/algolia/eslint-config-algolia#rules

One of the not so obvious rule is the semi colon remove, I was convinced
by https://github.com/feross/standard#rules and
https://github.com/yyx990803/semi#but-semicolons-are-required

I assume there are some rules like spacing aroud parentheses and
brackets that will be a big change for you @bobylito (and it's your
project so I am dropping your name here!)

But we need to try to find a style that is both OUR style and also will
not give our future open source contributors too much WTF moments when
encountering our style.

On this subject, http://sideeffect.kr/popularconvention/#javascript is a
good data based source.

We can discuss, add, remove, edit the rules as long as we do not discuss them
every week!

Other projects:

- Original feross/standard:
After trying it I discovered some rules were not strict enough and
would lead to again style discussions. Has some sort of formatter but
not very good. Based on eslint.

Ref:
- https://github.com/feross/standard/issues/180

- JSCS:
I tried JSCS: good but only oriented on style, no rules on unused
variables or ways to prevent bugs like eslint does. Has some sort of
formatter, but not very good.

Ultimately eslint will have a good formatter, in the meantime we will
all switch to some new coding style, for good :-)?

fixes #128

vvo commented:
We may have more luck in auto formatting with https://github.com/millermedeiros/esformatter and https://github.com/millermedeiros/esformatter/wiki/Plugins

bobylito commented on README.md:
```diff
@@ -62,19 +62,19 @@ helper.search();
 [See the examples in action](http://algolia.github.io/algoliasearch-helper-js/)
 
-### Use with NPM
+### With NPM
```
Just as a side note, maybe that's not the right PR for that ;)

vvo replied:
Ok will revert this

bobylito replied:
But that's not a big deal ;)

bobylito commented on examples/instantsearch+helper.html:
```diff
@@ -43,13 +44,13 @@
-          helper.setQuery( $inputfield.val() ).search();
+        $inputfield.keyup(function(e){
+          helper.setQuery($inputfield.val()).search();
         } );
```
Why do we have space at the end of a pair of parentheses and not at the beginning?

pixelastic replied:
We shouldn't have at the end either

vvo replied:
Yes that's a fail, I will check the rules to have no space at the end too.

vvo replied:
Ah, I know why, it's an html file so I will have to reformat those too, thx

bobylito commented on src/SearchParameters/RefinementList.js:
```diff
@@ -118,19 +121,19 @@ var lib = {
 
     return containsRefinements &&
-           refinementList[ attribute ].indexOf( refinementValueAsString ) !== -1;
+      refinementList[attribute].indexOf(refinementValueAsString) !== -1
```
Is that in the rules?

vvo replied:
Yes that's http://eslint.org/docs/rules/array-bracket-spacing and http://eslint.org/docs/rules/space-in-parens

bobylito replied:
OK not the square brackets :) but the alignements of the returned value

vvo replied:
Seems not to be enforced by any rule in eslint.

The only added value is that you do not have to hit multiple times <TAB> to align nor you have to precisely space with one more space keypress since it's an even spacing align here (11 spaces).

I think it got rewritten by [standard-format](https://github.com/maxogden/standard-format) when I first used it. standard-format uses esformatter and esformatter has an option for this kind of alignment rule.

I think I will create an esformatter algolia config so that we can easily reformat our code.

bobylito replied:
but it pleases my eyes ;) But I agree that's not so common, and we're looking for commonish stuff here so feel free to change it back or not ;)

@pixelastic started one :)

bobylito commented:
Imho your PR shouldn't contain the fixes for the CI...

vvo commented:
> Imho your PR shouldn't contain the fixes for the CI...

Agree. I only added them here in thinking that if we merge this then I would not have to re apply the rules to the test files, I will revert the commit and submit a new PR targeting this current PR

bobylito commented:
Ok so I went through the whole PR. Personally speaking I don't feel it actually matches "my style" visually :). The points that actually made me thinking were : 
 - the semicolon
 - the else / else if statements on the same line that the previous if / else if block end
 - the single quote

But really I shouldn't be cry baby here, that's awesome to get a common style :) :+1:

vvo commented:
On the semicolon side that's also a big change for me but I do have the feeling that in one month if we apply it everywhere we code we will love it!

We only need to make sure we do add the semicolon when we document things like on the website. On adding/removing semicolon easily on a whole project: https://www.npmjs.com/package/semi

I know it's an even bigger style change for you (more changes) and I appreciate you still find it awesome that we aim at a common style!

Gonna make some changes we talked and will ping again here

pixelastic commented:
Yep, no semi-colon is a big change for me, but still in favor of any unified codestyle. There is also a package to convert [double quotes to single quotes](https://github.com/sindresorhus/to-single-quotes) if you ever need it.

vvo commented:
> Imho your PR shouldn't contain the fixes for the CI...

Fixed, merge #135 before this one and we're good

vvo commented:
Reverted semi colons, fixed html examples. Now gonna write the esformatter for algolia and put it here also as a way to reformat easily our code.

Then will do the same for client-js.

bobylito commented:
<img src=http://i.giphy.com/pCO5tKdP22RC8.gif>

bobylito commented:
It fails on Iphone :/

vvo commented:
Success!
